### ISSUE TITLE
Missing drugs when updating to ChEMBL 29

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/platform#1759]
**Context**
The comparison of the numbers between the drug index based on ChEMBL 28 and ChEMBL 29 shows a decrease in the number of unique drugs (13 185 vs 12594). ChEMBL 29 represents an overall increase over the previous version, which indicates that there is something wrong with the way we generate the drug index.

There are 984 ChEMBL IDs that went missing between versions. Out of these, **968** of them are present in the dataset we produce out of the ChEMBL molecule dump (`gs://open-targets-pre-data-releases/21.09.1/input/annotation-files/chembl/chembl_29_molecule-2021-09-16.jsonl`). The other 16 have been removed by ChEMBL.

Here is the dataset with the offending molecules:
[https://app.zenhub.com/files/143733948/8fa001d8-0826-42a6-80ca-327acdfb59b0/download](https://app.zenhub.com/files/143733948/8fa001d8-0826-42a6-80ca-327acdfb59b0/download)

**Is it a pipeline issue?**
The drug step has suffered minor changes since the last release:
- There were some changes that needed to be made in PIS to generate the ChEMBL target dataset. I don't think this is the root cause as both versions of the datasets are identical.
- Some [minor improvements in the Drug step](https://github.com/opentargets/platform-etl-backend/commits/master/src/main/scala/io/opentargets/etl/backend/drug) that I don't see how they could cause the difference. 

**Is it a data issue?**
As @JarrodBaker pointed out, the requisites for a drug to be included are:
1. have an associated indication
2. have the MoA annotated
3. have a cross-reference to DrugBank.

I have compared them by creating a `hasIndication, hadIndication, hasMoa, hadMoa `columns.
- There are only 5 records whose indications have gone missing:
```
[Row(id='CHEMBL288376'),
 Row(id='CHEMBL4594270'),
 Row(id='CHEMBL4594358'),
 Row(id='CHEMBL253144'),
 Row(id='CHEMBL4298194')]
```
- And one molecule when MoAs are considered:
```
[Row(id='CHEMBL4298088')]
```
The DrugBank xrefs have not changed as these have been collected from the same input file(`gs://open-targets-pre-data-releases/21.09.1/input/annotation-files/drugbank-2020-06-01.csv.gz`).

I don't think we have changed how MoA and indications are inherited since 21.06, so I don't see how the definition of drug could account for the change. 

I'll keep digging, but any suggestion is welcome  @d0choa @DSuveges @mkarmona  @JarrodBaker @cmalangone

**Comment by d0choa** [comment]
I think the lack of molecules is explained by the work done to align with ChEMBL parent/child propagation of indications and MoAs #1590

This change (which I don’t think went live in 21.06) implied fewer indications in child molecules as a result of not propagating from parents to children anymore. This change will imply that a lot of the child molecules that had an indication in the past, don't have it anymore. As a result, many molecules won't fulfil the rule for inclusion in the platform (indication | MoA | Drugbank ID).

The next confirms that most of the molecules lost in the above list were indeed child molecules:

```
❯ grep -c parent problematic_chembl_ids.json
963
❯ grep -c id problematic_chembl_ids.json
968
```

**Comment by mkarmona** [comment]
@ireneisdoomed really well described issue :clap:

**Comment by ireneisdoomed** [comment]
@d0choa I kind of assumed these changes went live in 21.06 as they were merged mid-June and the indexes were uploaded to the release buckets on June 27th. It would be very nice to confirm it as that would definitely explain things.

**Comment by d0choa** [comment]
I retract my own words. I think there might be some bug related to #1590 that might be causing the disappearance of some MoAs and therefore some molecules.  I'll dig a bit further

**Comment by mkarmona** [comment]
@d0choa these are good examples to see what is failing here.

**Comment by mkarmona** [comment]
Looking closely at the two examples @d0choa wrote above `CHEMBL3094198` and `CHEMBL2104871` it seems their parent molecules are the ones holding the MOA data.
The case `CHEMBL3094198` does not hold any MOA but its parent does as I show below
```
❱ cat chembl_29_molecule-2021-09-16.jsonl | grep CHEMBL3094198 | jqless
{
  "molecule_chembl_id": "CHEMBL3094198",
  "molecule_hierarchy": {
    "parent_chembl_id": "CHEMBL1431",
    "molecule_chembl_id": "CHEMBL3094198"
  },
  "molecule_type": "Small molecule",
...
```
Let's check the parent's MOAs
```
◆ cat chembl_29_mechanism-2021-08-20.jsonl | grep 'CHEMBL1431'
{"molecule_chembl_id": "CHEMBL1703", "target_chembl_id": "CHEMBL3391681", "action_type": "INHIBITOR", "mechanism_of_action": "Mitochondrial glycerol-3-phosphate dehydrogenase inhibitor", "mechanism_refs": [{"ref_id": "24847880", "ref_url": "http://europepmc.org/abstract/MED/24847880", "ref_type": "PubMed"}], "record_id": 1344532, "parent_molecule_chembl_id": "CHEMBL1431", "_metadata": {"parent_molecule_chembl_id": "CHEMBL1431", "all_molecule_chembl_ids": ["CHEMBL1703", "CHEMBL1431"], "should_appear_in_browser": false}}
{"molecule_chembl_id": "CHEMBL1703", "target_chembl_id": "CHEMBL2363065", "action_type": "INHIBITOR", "mechanism_of_action": "Mitochondrial complex I (NADH dehydrogenase) inhibitor", "mechanism_refs": [{"ref_id": "23277337", "ref_url": "http://europepmc.org/abstract/MED/23277337", "ref_type": "PubMed"}, {"ref_id": "23358356", "ref_url": "http://europepmc.org/abstract/MED/23358356", "ref_type": "PubMed"}, {"ref_id": "25017630", "ref_url": "http://europepmc.org/abstract/MED/25017630", "ref_type": "PubMed"}], "record_id": 1344532, "parent_molecule_chembl_id": "CHEMBL1431", "_metadata": {"parent_molecule_chembl_id": "CHEMBL1431", "all_molecule_chembl_ids": ["CHEMBL1703", "CHEMBL1431"], "should_appear_in_browser": false}}
```
Checking these target Chembl IDs
```
◆ cat chembl_29_target-2021-09-16.jsonl | grep 'CHEMBL3391681'|jq -r '.target_components[].accession'
P43304
```
And a further look at our target index we can double-check if the accession is correct
```
@ T.filter($"symbols" === "P43304").show 
+---------------+--------------+-------+
|             id|approvedSymbol|symbols|
+---------------+--------------+-------+
|ENSG00000115159|          GPD2| P43304|
+---------------+--------------+-------+
```
And taking some random accession from the other target and resolving back from our index
```
@ T.filter($"symbols" === "P03923").show 
+---------------+--------------+-------+
|             id|approvedSymbol|symbols|
+---------------+--------------+-------+
|ENSG00000198695|        MT-ND6| P03923|
+---------------+--------------+-------+
```

Repeating the same procedure with the other molecule shows that the accession id is not human therefore it is contained neither in the UniProt human file nor the target index 
```
@ T.filter($"symbols" === "P54300").show 
+---+--------------+-------+
| id|approvedSymbol|symbols|
+---+--------------+-------+
+---+--------------+-------+
```
So the most probable hypothesis to pose is whether **Chembl is projecting some parents' MOAs onto their children molecules** @ireneisdoomed @JarrodBaker. A further question is whether this is a systematic approach from them or an uncovered bug as you can see in their JSON a field called `"should_appear_in_browser": false` which could mean nothing tbh.

**Comment by d0choa** [comment]
In conclusion:

- Our codebase is working as specified. MoAs and Indications are only propagated child -> parent (#1590) and molecules are only included if (Indication | MoA | DrugBank)
- The majority of missing molecules are child molecules before (#1590) kicked in. Because they lost the inherited MoA they now don't qualify as valid molecules.
- ChEMBL web does seem to propagate MoAs in the wrong direction parent -> child in the [offending molecules](https://app.zenhub.com/files/143733948/8fa001d8-0826-42a6-80ca-327acdfb59b0/download). This propagation happens at their API level and it goes against the propagation rule they recommended.

Actions:
- @ireneisdoomed to confirm with ChEMBL why their site does not comply with the rules discussed on the [3/6/21 meeting](https://github.com/opentargets/platform/issues/1590#issue-911440133)

**Comment by JarrodBaker** [comment]
What a ticket! I was heading in the same direction as you @d0choa . When I look at mechanisms of action in 21_06 data the missing molecules are found in the `chemblIds` field along with their parent, but in 21_09 they are disappearing.

**Comment by ireneisdoomed** [comment]
After a conversation with Juan and Patricia from ChEMBL, everything looks like we are safe and this drop makes sense.
- For the example we were discussing, `CHEMBL3094198`, it is not that the MoA is being inherited from the parent, but from another salt. 
- This is a bug in ChEMBL's logic that they will be resolving in their UI. When we had the discussion in June (#1590), the approach they recommended was not entirely implemented on their side. These molecules that we are dropping now shouldn't have been there in the first place.
- I have provided them with a list of these compounds for them to test.
- For the other 5 molecules which are not children and hence do not follow this pattern, they are gone because they used to be in the MoA/indication dataset. Patricia said that this could be caused by that same molecule being represented with another CHEMBLID. I have also provided these and she'll look everything makes sense.

Related to the last point, one interesting thing I learnt was that every InChiKey is represented by a single CHEMBLID. Therefore, for a given molecule, if the known molecular information changes they will create another ID to represent that molecule (and hopefully deprecate the obsolete one).

**I consider this ticket as closed with no actions pending to be done on our side.**
